src/transposition.cpp

```diff
@@ -281,9 +281,21 @@ void Transposer::SetTransposition(int transVal)
 // Use a string to set the interval class in the current base system.  For example,
 //  "+M2" means up a major second, which is the integer 6 in base-40.
 
-void Transposer::SetTransposition(const std::string &transString)
+bool Transposer::SetTransposition(const std::string &transString)
 {
     m_transpose = GetIntervalClass(transString);
+    return m_transpose != INVALID_INTERVAL_CLASS;
+}
+
+bool Transposer::SetTransposition(const TransPitch &fromPitch, const std::string &toString)
+{
+    TransPitch toPitch;
+    if (GetKeyTonic(toString, toPitch)) {
+        // TODO: Determine proper octave offset.
```

Comment:
Thinking about the logic for correcting the octave,

assuming fromPitch is octave 0 (writing **F0** as *F*)

* Transposing from *F* to `G`: no change needed (**G0** is the closest G)
* Transposing from *F* to `+G`: subtract one octave (the first `+` determines the direction) (for all other `+++G`s, also subtract one octave)
* Transposing from *F* to `-G`: no change needed (**G-1** is the closest downward G)

* Transposing from *G* to `F`: no change needed (**F0** is the closest F)
* Transposing from *G* to `+F`: no change needed (**F1** is the closest upward F)
* Transposing from *G* to `-F`: add one octave (so it gives the same result as `F`)

* Transposing from *C* to `B`: Add one octave (**B1** is the closest B)
* Transposing from *C* to `+B`: Do nothing (already was giving **B1**)
* Transposing from *C* to `-B`: Add one octave (**B0** is the closest downward B)

* Transposing from *B* to `C`: Subtract one octave (**C-1** is the closest C)
* Transposing from *B* to `+C`: Subtract one octave (**C0** is the closest upward C)
* Transposing from *B* to `-C`: do nothing (**C-1** is already the closest downward C)

Probably would be good to think of a way to simplify the logic here without needing 12 cases

Reply:
I don't understand this case:

> Transposing from F to +G: subtract one octave (the first + determines the direction) (for all other +++Gs, also subtract one octave)

Here is how I am interpreting the pluses and the output octave for each cases:

F to G is a major second, so F0 to G0.
F to +G is a major second, so F0 to G0.
F to ++G is a major second and an octave, so F0 to G1.
F to +++G is a major second and two octaves, to F0 to G2.

Reply:
By `subtract one octave`, I guess I mean compared to the offset that GetKeyTonic gives:

`+G` = **G1**, so we need to subtract an octave from this so it become **G0**, as we expect.

Reply:
Maybe a better way of saying it (that might help with the logic):

A transposition with **_n_** plus `+` or minus `-` signs should never be more than **_n_** octaves away.

Reply:
And a transposition with **_0_** plus or minus signs should never be more than 1/2 an octave away.

Reply:
`+G` cannot be interpreted as an absolute case since it is relative to the starting key.  `+G` will always be up a particular interval which is an octave or less.

The algorithm that I am thinking when there are pluses present:

From the key of F to `+G` for example:

(1) Start with F0 and G0 (both pitches in the 0th octave) and convert to both to integer pitches.
(2) if G0 is a smaller number than F0, add m_base to G0 
       (this is not necessary for the example case, but in the general case this will be possible such as when transposing from G up to F).
(3) for each left-over `+` sign, also add m_base to the target key's pitch interval.
(4) calculate the interval by subtracting F0 from the octave-adjusted G0 pitch.

Here is the reverse upward case: from G to +F:
(1) start with G0 and F0.
(2) F0 is smaller than G0, so add an octave to move it to F1.
(3) There are no more plus signs, so F1 no longer needs to be adjusted.
(4) Calculate the interval class: F1 - G0, which is up a minor seventh.


For the forced down cases:

For example from G to `-F`:

(1) Start with G0 and F0
(2) F0 is  less than G0, so no octave adjustment is necessary
(3) there are no extra `-` signs, so no further octave adjustment
(4) F0 - G0 = down a major second

Transpose from F major down to G major:

(1) Start with F0 and G0
(2) G0 is greater than F0, so change to G-1
(3) There are not extra `-` signs to no extra octave adjustment
(4) G-1 - F0 = down a minor seventh

------------

For the case where there is neither `+` nor  `-`, then the algorithm has to check which direction has the smallest absolute interval:

from key X to key Y:

(1) calculate the interval pitches for X0 and Y0
(2) If (Y0 - X0) is positive then compare to abs(Y0 - X-1) and choose whichever of X0 and X-1 gives the smallest value.
(2) If (Y0 - X0) is negative, then compare to abs(Y0 - X1) and choose whichever of X0 and X1 gives the smallest value.


So there are three cases to implement: When there are pluses, when there are minuses, and when there are neither pluses or minuses.  The first two cases could be merged into the same code, but it is probably clearer to keep them separate.

Reply:
I implemented it in c1b7f8e745989c13e8ec7f20a0ec8c0a336f1f31

I'm using four cases (I've split the case with neither pluses nor minuses into two) It could be simplified into an `if` and an `else if`, but right now I think it's pretty understandable.